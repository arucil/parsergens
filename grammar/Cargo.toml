[package]
name = "grammar"
version = "0.1.0"
authors = ["amlo <xplzjwz@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lalrpop-util = "0.19.0"
bit-set = "0.5.2"
indexmap = "1.5.1"
bimap = "0.5.2"
either = "1.6.0"
codespan-reporting = "0.9.5"
bitvec = "0.18.1"

[build-dependencies]
lalrpop = "0.19.0"

[dev-dependencies]
insta = "0.16.1"
pretty_assertions = "0.6.1"
