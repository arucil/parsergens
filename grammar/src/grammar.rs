use std::ops::Range;
use std::borrow::Borrow;
use super::lexer::{Lexer, TokenId};
use crate::{Set, BiMap, Map};

pub use lower::*;

pub mod lower;

#[derive(Debug)]
pub struct Grammar {
  /// productions of the same nonterminals are consecutive.
  pub rules: Vec<Rule>,
  /// starting non-terminals
  pub start_nts: Set<NonterminalId>,
  /// non-terminals
  pub nts: BiMap<NonterminalId, String>,
  pub nt_prods: Map<NonterminalId, Range<usize>>,
  pub lexer: Lexer,
}

#[derive(Debug)]
pub struct Rule {
  pub nt: NonterminalId,
  pub items: Vec<Item>,
}

#[derive(Debug)]
pub enum Item {
  Nonterminal(NonterminalId),
  Token(TokenId),
  Optional(Vec<Item>),
  Many(Vec<Item>),
  Many1(Vec<Item>),
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
pub struct NonterminalId(u32);

impl NonterminalId {
  pub fn id(&self) -> u32 {
    self.0
  }
}

impl Borrow<u32> for NonterminalId {
  fn borrow(&self) -> &u32 {
    &self.0
  }
}

#[derive(Default)]
pub struct NonterminalIdGen(u32);

impl NonterminalIdGen {
  pub fn gen(&mut self) -> NonterminalId {
    let i = self.0;
    self.0 += 1;
    NonterminalId(i)
  }

  pub fn from(start: u32) -> Self {
    Self(start + 1)
  }
}

impl Grammar {
  pub fn lower(self) -> LoweredGrammar {
    lower::lower(self)
  }
}