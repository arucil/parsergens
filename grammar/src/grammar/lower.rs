use std::ops::Range;
use crate::{
  Grammar, NonterminalIdGen, Item, Lexer, TokenId, NonterminalId
};
use crate::{Set, BiMap, Map};

#[derive(Debug)]
pub struct LoweredGrammar {
  pub prods: Vec<Production>,
  pub start_nts: Set<NonterminalId>,
  pub nts: BiMap<NonterminalId, String>,
  pub nt_metas: Map<NonterminalId, LoweredNonterminalMetadata>,
  pub lexer: Option<Lexer>,
  pub tokens: BiMap<TokenId, String>,
  pub user_code: Vec<String>,
}

#[derive(Debug)]
pub struct Production {
  pub nt: NonterminalId,
  pub kind: ProductionKind,
  pub symbols: Vec<Symbol>,
  pub action: Option<String>,
}

#[derive(Debug)]
pub struct LoweredNonterminalMetadata {
  pub range: Range<usize>,
  pub ty: Option<String>,
  pub kind: NonterminalKind,
}

#[derive(Debug, Clone, Copy)]
pub enum NonterminalKind {
  User,
  Repetition,
  Optional,
}

#[derive(Debug, Clone, Copy)]
pub enum ProductionKind {
  Ordinary,
  RepetitionFirst,
  RepetitionRest,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Symbol {
  Nonterminal(NonterminalId),
  Token(TokenId),
}

pub(super) fn lower(grammar: Grammar) -> LoweredGrammar {
  let max_nt_id = grammar.nts
    .left_values()
    .map(|x| x.id())
    .max()
    .unwrap();
  let mut nt_id_gen = NonterminalIdGen::from(max_nt_id);

  let mut lowered = LoweredGrammar {
    prods: vec![],
    start_nts: grammar.start_nts,
    nts: grammar.nts,
    nt_metas: Map::new(),
    lexer: grammar.lexer,
    tokens: grammar.tokens,
    user_code: grammar.user_code,
  };

  for (nt, meta) in grammar.nt_metas {
    let mut rules = vec![];

    for rule in &grammar.rules[meta.range] {
      rules.push((
        lower_items(&rule.items, &mut lowered, &mut nt_id_gen),
        rule.action.clone()
      ));
    }

    let start = lowered.prods.len();

    for (symbols, action) in rules {
      lowered.prods.push(Production {
        nt,
        kind: ProductionKind::Ordinary,
        symbols,
        action,
      });
    }

    let end = lowered.prods.len();

    let meta = LoweredNonterminalMetadata {
      range: start..end,
      ty: meta.ty,
      kind: NonterminalKind::User,
    };

    lowered.nt_metas.insert(nt, meta);
  }

  lowered
}

fn lower_items(
  items: &[Item],
  lowered: &mut LoweredGrammar,
  nt_id_gen: &mut NonterminalIdGen,
) -> Vec<Symbol> {
  items.iter().map(|item| {
    match item {
      Item::Nonterminal(nt) => Symbol::Nonterminal(*nt),
      Item::Token(token) => Symbol::Token(*token),
      Item::Optional(items) => {
        let symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '?');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::Ordinary,
          symbols: vec![],
          action: None,
        });
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::Ordinary,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Optional,
        });

        Symbol::Nonterminal(nt)
      }
      Item::Many(items) => {
        let mut symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '*');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionFirst,
          symbols: vec![],
          action: None,
        });

        symbols.insert(0, Symbol::Nonterminal(nt));
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionRest,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Repetition,
        });

        Symbol::Nonterminal(nt)
      }
      Item::Many1(items) => {
        let mut symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '+');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionFirst,
          symbols: symbols.clone(),
          action: None,
        });

        symbols.insert(0, Symbol::Nonterminal(nt));
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionRest,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Repetition,
        });

        Symbol::Nonterminal(nt)
      }
    }
  }).collect()
}

fn make_production_name(
  symbols: &[Symbol],
  lowered: &LoweredGrammar,
  suffix: char
) -> String {
  let mut buf = "(".to_owned();
  let mut space = false;

  for symbol in symbols {
    if space {
      buf.push(' ');
    }

    match symbol {
      Symbol::Nonterminal(nt) => {
        buf.push_str(lowered.nts.get_by_left(nt).unwrap());
      }
      Symbol::Token(token) => {
        buf.push_str(lowered.tokens.get_by_left(token).unwrap());
      }
    }

    space = true;
  }

  buf.push(')');
  buf.push(suffix);

  buf
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn lower_repetition() {
    let grammar = crate::build(r#"
%token A "a"

%start top

top = A (atom (A)? )* A
  | (A atom)+

atom = A
  | ()
    "#).unwrap();

    let lowered = lower(grammar);

    assert_debug_snapshot!(lowered);
  }
}