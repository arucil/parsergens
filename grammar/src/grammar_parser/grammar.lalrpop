use lalrpop_util::ParseError;
use super::ast::*;
use super::lex;
use super::UserParseError;
use super::regex;

#[LALR]
grammar<'input>(input: &'input str);

pub Document: Grammar = (Separator <decl>)*
  => {
    <>.into_iter()
      .filter_map(|x| x)
      .collect()
  };

decl: Option<Spanned<Decl>> = {
  <l:@L> <decl:rule_decl> <r:@L> => Some(Spanned((l, r), Decl::Rule(decl))),
  <l:@L> <decl:start_decl> <r:@L> => Some(Spanned((l, r), Decl::Start(decl))),
  <l:@L> <decl:token_decl> <r:@L> => Some(Spanned((l, r), Decl::Token(decl))),
  <l:@L> <decl:ext_token_decl> <r:@L> => Some(Spanned((l, r), Decl::ExtToken(decl))),
  <l:@L> <decl:skip_decl> <r:@L> => Some(Spanned((l, r), Decl::Skip(decl))),
  <l:@L> <decl:user_decl> <r:@L> => Some(Spanned((l, r), Decl::User(decl))),
  <l:@L> <decl:assoc_decl> <r:@L> => Some(Spanned((l, r), Decl::Assoc(decl))),
  <l:@L> <decl:state_decl> <r:@L> => Some(Spanned((l, r), Decl::State(decl))),
};

rule_decl: RuleDecl =
  <name:ident> <ty:(Colon <code_block>)?> Assign <alt1:production> <mut alts:(Or <production>)*>
  =>? {
    alts.insert(0, alt1);
    if alts.iter().any(|alt| alt.1.action.is_some()) && ty.is_none() {
      Err(ParseError::User {
        error: UserParseError::MissingType(name.0),
      })
    } else {
      Ok(RuleDecl {
        name,
        ty,
        alts,
      })
    }
  }; 

production: Spanned<RuleAlt> = {
  <l:@L> <terms:term+> <prec:(Percent Prec <ident>)?> <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Terms(terms),
        prec,
        action,
      }
    ),
  <l:@L> epsilon <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Epsilon,
        prec: None,
        action,
      }
    ),
};

epsilon: () = LParen RParen;

term: Term = {
  ident => Term::Symbol(<>),
  LParen <terms:term+> RParen QuestionMark => Term::Optional(terms),
  LParen <terms:term+> RParen Asterisk => Term::Many(terms),
  LParen <terms:term+> RParen Plus => Term::Many1(terms),
}

start_decl: StartDecl = Percent Start <name:ident>
  => {
    StartDecl {
      name,
    }
  };

token_decl: TokenDecl = Percent Token <name:ident> <pattern:token_pattern>
  => {
    TokenDecl {
      name,
      pattern,
    }
  };

ext_token_decl: ExternalTokenDecl = Percent Token <name:ident>
  => {
    ExternalTokenDecl {
      name,
    }
  };

skip_decl: SkipDecl = Percent Skip <pattern:token_pattern>
  => {
    SkipDecl {
      pattern,
    }
  };

user_decl: UserDecl = Percent User <code:code_block>
  => {
    UserDecl {
      code,
    }
  };

assoc_decl: AssocDecl = Percent <assoc:assoc> <names:ident+>
  => {
    AssocDecl {
      names,
      assoc,
    }
  };

assoc: Spanned<Assoc> = {
  <l:@L> LeftAssoc <r:@L> => Spanned((l, r), Assoc::LeftAssoc),
  <l:@L> RightAssoc <r:@L> => Spanned((l, r), Assoc::RightAssoc),
  <l:@L> NonAssoc <r:@L> => Spanned((l, r), Assoc::NonAssoc),
}

state_decl: StateDecl = Percent State <state:code_block>
  => {
    StateDecl {
      state,
    }
  };

token_pattern: TokenPattern = {
  <regex:regex> =>? regex::parse_regex(
    &regex.1[1..regex.1.len() - 1],
    regex.0 .0 + 1
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::Regex,
      source: regex,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error),
    }),
  <str:string> =>? regex::parse_string_literal(
    &str.1[1..str.1.len() - 1], str.0 .0
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::String,
      source: str,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error)
    }),
}

spanned<T>: Spanned<String> = <l:@L> <x:T> <r:@L> => Spanned((l, r), x.text.to_owned());

ident = spanned<Ident>;
string = spanned<String>;
regex = spanned<Regex>;
code_block: Spanned<String> = <spanned<CodeBlock>> =>
  Spanned(<>.0, <>.1.trim_start_matches('{').trim_end_matches('}').trim().to_owned());

extern {
  type Location = usize;
  type Error = UserParseError;

  enum lex::Token<'input> {
    Start => lex::Token { kind: lex::TokenKind::Start, .. },
    Token => lex::Token { kind: lex::TokenKind::Token, .. },
    Skip => lex::Token { kind: lex::TokenKind::Skip, .. },
    User => lex::Token { kind: lex::TokenKind::User, .. },
    LeftAssoc => lex::Token { kind: lex::TokenKind::LeftAssoc, .. },
    RightAssoc => lex::Token { kind: lex::TokenKind::RightAssoc, .. },
    NonAssoc => lex::Token { kind: lex::TokenKind::NonAssoc, .. },
    Prec => lex::Token { kind: lex::TokenKind::Prec, .. },
    State => lex::Token { kind: lex::TokenKind::State, .. },

    Ident => lex::Token { kind: lex::TokenKind::Ident, .. },
    Regex => lex::Token { kind: lex::TokenKind::Regex, .. },
    String => lex::Token { kind: lex::TokenKind::String, .. },
    CodeBlock => lex::Token { kind: lex::TokenKind::CodeBlock, .. },

    Percent => lex::Token { kind: lex::TokenKind::Percent, .. },
    Assign => lex::Token { kind: lex::TokenKind::Assign, .. },
    Or => lex::Token { kind: lex::TokenKind::Or, .. },
    LParen => lex::Token { kind: lex::TokenKind::LParen, .. },
    RParen => lex::Token { kind: lex::TokenKind::RParen, .. },
    Comma => lex::Token { kind: lex::TokenKind::Comma, .. },
    Colon => lex::Token { kind: lex::TokenKind::Colon, .. },
    Asterisk => lex::Token { kind: lex::TokenKind::Asterisk, .. },
    QuestionMark => lex::Token { kind: lex::TokenKind::QuestionMark, .. },
    Plus => lex::Token { kind: lex::TokenKind::Plus, .. },
    Arrow => lex::Token { kind: lex::TokenKind::Arrow, .. },

    Enter => lex::Token { kind: lex::TokenKind::Enter, .. },
    Leave => lex::Token { kind: lex::TokenKind::Leave, .. },
    Separator => lex::Token { kind: lex::TokenKind::Separator, .. },
  }
}