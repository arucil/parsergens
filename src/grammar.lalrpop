use super::ast::*;
use super::lex;

#[LALR]
grammar<'input>(input: &'input str);

pub document: Grammar = <decl*>
  => {
    <>.into_iter()
      .filter_map(|x| x)
      .collect()
  };

decl: Option<Spanned<Decl>> = {
  <l:@L> <decl:rule_decl> <r:@L> => Some(Spanned(l..r, Decl::Rule(decl))),
  <l:@L> <decl:start_decl> <r:@L> => Some(Spanned(l..r, Decl::Start(decl))),
  <l:@L> <decl:token_decl> <r:@L> => Some(Spanned(l..r, Decl::Token(decl))),
  Newline => None,
};

rule_decl: RuleDecl =
  <name:ident> Assign <alt1:production> Newline <mut alts:(Indent Or <production> Newline)*>
  => {
    alts.insert(0, alt1);
    RuleDecl {
      name,
      alts,
    }
  }; 

production: Spanned<RuleAlt> = {
  <l:@L> <terms:ident*> <r:@L> => Spanned(l..r, RuleAlt::Terms(terms)),
  <l:@L> Epsilon <r:@L> => Spanned(l..r, RuleAlt::Epsilon),
};

start_decl: StartDecl = Percent Start <name:ident> Newline
  => {
    StartDecl {
      name,
    }
  };

token_decl: TokenDecl = Percent Token <name:ident> <regex:regex>
  => {
    TokenDecl {
      name,
      regex,
    }
  };

spanned<T>: Spanned<String> = <l:@L> <x:T> <r:@L> => Spanned(l..r, x.text.to_owned());

ident = spanned<Ident>;
regex = spanned<Regex>;

extern {
  type Location = usize;
  type Error = lex::LexError;

  enum lex::Token<'input> {
    Ident => lex::Token { kind: lex::TokenKind::Ident, .. },
    Start => lex::Token { kind: lex::TokenKind::Start, .. },
    Token => lex::Token { kind: lex::TokenKind::Token, .. },
    Regex => lex::Token { kind: lex::TokenKind::Regex, .. },

    Percent => lex::Token { kind: lex::TokenKind::Percent, .. },
    Assign => lex::Token { kind: lex::TokenKind::Assign, .. },
    Or => lex::Token { kind: lex::TokenKind::Or, .. },
    Epsilon => lex::Token { kind: lex::TokenKind::Epsilon, .. },
    Comma => lex::Token { kind: lex::TokenKind::Comma, .. },

    Indent => lex::Token { kind: lex::TokenKind::Indent, .. },
    Newline => lex::Token { kind: lex::TokenKind::Newline, .. },
  }
}