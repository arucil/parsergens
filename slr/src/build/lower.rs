use std::ops::Range;
use grammar::{Grammar, NonterminalIdGen, Item};
use crate::{Set, BiMap, Map};

pub use grammar::{NonterminalId, Lexer, TokenId};

#[derive(Debug)]
pub struct LoweredGrammar {
  pub prods: Vec<Production>,
  pub start_nts: Set<NonterminalId>,
  pub nts: BiMap<NonterminalId, String>,
  pub nt_prods: Map<NonterminalId, Range<usize>>,
  pub lexer: Lexer,
}

#[derive(Debug)]
pub struct Production {
  pub nt: NonterminalId,
  pub action: ProductionAction,
  pub symbols: Vec<Symbol>,
}

#[derive(Debug)]
pub enum ProductionAction {
  None,
  RepetitionFirst,
  RepetitionRest,
}

#[derive(Clone, Debug)]
pub enum Symbol {
  Nonterminal(NonterminalId),
  Token(TokenId),
}

pub fn lower(grammar: Grammar) -> LoweredGrammar {
  let max_nt_id = grammar.nts
    .left_values()
    .map(|x| x.id())
    .max()
    .unwrap();
  let mut nt_id_gen = NonterminalIdGen::from(max_nt_id);

  let mut lowered = LoweredGrammar {
    prods: vec![],
    start_nts: grammar.start_nts,
    nts: grammar.nts,
    nt_prods: Map::new(),
    lexer: grammar.lexer,
  };

  for (nt, range) in grammar.nt_prods {
    let mut prod_symbols = vec![];

    for prod in &grammar.prods[range] {
      prod_symbols.push(
        lower_items(&prod.items, &mut lowered, &mut nt_id_gen));
    }

    let start = lowered.prods.len();

    for symbols in prod_symbols {
      lowered.prods.push(Production {
        nt,
        action: ProductionAction::None,
        symbols,
      });
    }

    let end = lowered.prods.len();

    lowered.nt_prods.insert(nt, start..end);
  }

  lowered
}

fn lower_items(
  items: &[Item],
  lowered: &mut LoweredGrammar,
  nt_id_gen: &mut NonterminalIdGen,
) -> Vec<Symbol> {
  items.iter().map(|item| {
    match item {
      Item::Nonterminal(nt) => Symbol::Nonterminal(*nt),
      Item::Token(token) => Symbol::Token(*token),
      Item::Optional(items) => {
        let symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '?');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          action: ProductionAction::None,
          symbols: vec![],
        });
        lowered.prods.push(Production {
          nt,
          action: ProductionAction::None,
          symbols,
        });

        let end = lowered.prods.len();

        lowered.nt_prods.insert(nt, start..end);

        Symbol::Nonterminal(nt)
      }
      Item::Many(items) => {
        let mut symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '*');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          action: ProductionAction::RepetitionFirst,
          symbols: vec![],
        });

        symbols.insert(0, Symbol::Nonterminal(nt));
        lowered.prods.push(Production {
          nt,
          action: ProductionAction::RepetitionRest,
          symbols,
        });

        let end = lowered.prods.len();

        lowered.nt_prods.insert(nt, start..end);

        Symbol::Nonterminal(nt)
      }
      Item::Many1(items) => {
        let mut symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '+');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          action: ProductionAction::RepetitionFirst,
          symbols: symbols.clone(),
        });

        symbols.insert(0, Symbol::Nonterminal(nt));
        lowered.prods.push(Production {
          nt,
          action: ProductionAction::RepetitionRest,
          symbols,
        });

        let end = lowered.prods.len();

        lowered.nt_prods.insert(nt, start..end);

        Symbol::Nonterminal(nt)
      }
    }
  }).collect()
}

fn make_production_name(
  symbols: &[Symbol],
  lowered: &LoweredGrammar,
  suffix: char
) -> String {
  let mut buf = "(".to_owned();
  let mut space = false;

  for symbol in symbols {
    if space {
      buf.push(' ');
    }

    match symbol {
      Symbol::Nonterminal(nt) => {
        buf.push_str(lowered.nts.get_by_left(nt).unwrap());
      }
      Symbol::Token(token) => {
        buf.push_str(lowered.lexer.tokens.get_by_left(token).unwrap());
      }
    }

    space = true;
  }

  buf.push(')');
  buf.push(suffix);

  buf
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn lower_repetition() {
    let grammar = grammar::build(r#"
%token A "a"

%start top

top = A (atom (A)? )* A
  | (A atom)+

atom = A
  | ()
    "#).unwrap();

    let lowered = lower(grammar);

    assert_debug_snapshot!(lowered);
  }
}