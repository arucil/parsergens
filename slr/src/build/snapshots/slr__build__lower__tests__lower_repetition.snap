---
source: slr/src/build/lower.rs
expression: lowered
---
LoweredGrammar {
    prods: [
        Production {
            nt: NonterminalId(
                2,
            ),
            action: None,
            symbols: [],
        },
        Production {
            nt: NonterminalId(
                2,
            ),
            action: None,
            symbols: [
                Token(
                    TokenId(
                        0,
                    ),
                ),
            ],
        },
        Production {
            nt: NonterminalId(
                3,
            ),
            action: RepetitionFirst,
            symbols: [],
        },
        Production {
            nt: NonterminalId(
                3,
            ),
            action: RepetitionRest,
            symbols: [
                Nonterminal(
                    NonterminalId(
                        3,
                    ),
                ),
                Nonterminal(
                    NonterminalId(
                        1,
                    ),
                ),
                Nonterminal(
                    NonterminalId(
                        2,
                    ),
                ),
            ],
        },
        Production {
            nt: NonterminalId(
                4,
            ),
            action: RepetitionFirst,
            symbols: [
                Token(
                    TokenId(
                        0,
                    ),
                ),
                Nonterminal(
                    NonterminalId(
                        1,
                    ),
                ),
            ],
        },
        Production {
            nt: NonterminalId(
                4,
            ),
            action: RepetitionRest,
            symbols: [
                Nonterminal(
                    NonterminalId(
                        4,
                    ),
                ),
                Token(
                    TokenId(
                        0,
                    ),
                ),
                Nonterminal(
                    NonterminalId(
                        1,
                    ),
                ),
            ],
        },
        Production {
            nt: NonterminalId(
                0,
            ),
            action: None,
            symbols: [
                Token(
                    TokenId(
                        0,
                    ),
                ),
                Nonterminal(
                    NonterminalId(
                        3,
                    ),
                ),
                Token(
                    TokenId(
                        0,
                    ),
                ),
            ],
        },
        Production {
            nt: NonterminalId(
                0,
            ),
            action: None,
            symbols: [
                Nonterminal(
                    NonterminalId(
                        4,
                    ),
                ),
            ],
        },
        Production {
            nt: NonterminalId(
                1,
            ),
            action: None,
            symbols: [
                Token(
                    TokenId(
                        0,
                    ),
                ),
            ],
        },
        Production {
            nt: NonterminalId(
                1,
            ),
            action: None,
            symbols: [],
        },
    ],
    start_nts: {
        NonterminalId(
            0,
        ),
    },
    nts: {NonterminalId(0) <> "top", NonterminalId(1) <> "atom", NonterminalId(2) <> "(A)?", NonterminalId(3) <> "(atom (A)?)*", NonterminalId(4) <> "(A atom)+"},
    nt_prods: {
        NonterminalId(
            2,
        ): 0..2,
        NonterminalId(
            3,
        ): 2..4,
        NonterminalId(
            4,
        ): 4..6,
        NonterminalId(
            0,
        ): 6..8,
        NonterminalId(
            1,
        ): 8..10,
    },
    lexer: Lexer {
        dfa: TabularDfa {
            start: 0,
            state_base: [
                0,
                0,
                0,
            ],
            transitions: [
                (
                    0,
                    0,
                ),
                (
                    0,
                    2,
                ),
            ],
            accept_states: {
                State(1): TokenId(
                    0,
                ),
            },
        },
        char_intervals: [
            0,
            97,
            98,
        ],
        tokens: {TokenId(0) <> "A"},
    },
}
